### PR TITLE
mobile: Fix onDefaultNetworkChanged implementation on Android

### PR DESCRIPTION
**Pull request by fredyw**
This PR updates the implementation of `onDefaultNetworkChanged` for Android to only call the `onDefaultNetworkChanged` callback when there is a change in the network / transport type, such as when switching from WiFi to cellular, etc and not every network capability change.

Risk Level: medium
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: mobile (android)

### CODE REVIEW

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:98]: nits: update the cached value only when hasTransport(transportType) returns false?

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:83]: Only after Build.VERSION_CODES.O according to the documentation. Can we make sure this callback is only registered for version O+?

File: mobile/library/common/internal_engine.cc
Comment (danzh2010) [inline mobile/library/common/internal_engine.cc:288]: IMO, no need to rename it. We eventually want this callback to be triggered not just by network type, but also the network itself.

**Review by danzh2010 — COMMENTED** [review]
Thanks for the quick fix!

File: mobile/library/common/internal_engine.cc
Comment (fredyw) [reply mobile/library/common/internal_engine.cc:288]: I think it's probably better to have a different callback for that. This makes it similar to how it's implemented in Chromium: https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=974

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:83]: The `NetworkCallback` API was added in API level 21 but the guarantee is only starting with `Build.VERSION_CODES.O`.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:92]: Doh.. I'm dumb :( Thanks.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:98]: Done.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (abeyad) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:107]: does networkCapabilities only contain one transport type?  If not, is it correct to only return the first transport type encountered here?

File: mobile/library/common/internal_engine.cc
Comment (danzh2010) [reply mobile/library/common/internal_engine.cc:288]: The caller of the chromium interface also calls it when there is non-type change though: https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=1390-1396;drc=8e948282d37c0e119e3102236878d6f4d5052c16;bpv=1;bpt=1

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:107]: In the now deprecated API that was used in the previous code, there can only be a single network type: https://developer.android.com/reference/android/net/NetworkInfo#getType()

With the new API it's possible to have multiple transport types but our Envoy Mobile API doesn't even support that, i.e. `NetworkType`. We should probably rethink about the whole `NetworkType` but for now let's keep it simple.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:83]: In versions under O, this logic won't work right? In Chromium this callback is only registered for version R+.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:93]: Can you move this into onDefaultNetworkTypeChanged()?

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [inline mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java:115]: Should TRANSPORT_WIFI_AWARE also be regarded as WLAN https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=681;drc=8e948282d37c0e119e3102236878d6f4d5052c16;bpv=1;bpt=1?

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: Done.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: Let me remove the comment. I don't think it makes sense since we don't do anything with `onAvailable` at the moment. The `onCapabilitiesChanged` callback itself is available since API 21 (https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onCapabilitiesChanged(android.net.Network,%20android.net.NetworkCapabilities)), which is the minimum SDK version that we use: https://github.com/envoyproxy/envoy/blob/1130fca31d99f9c7ceff7fd0a1d2c4ee8ca25064/mobile/bazel/android_artifacts.bzl#L311

The Chromium code probably does something more that requires a higher API.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: Good point. Done.

File: mobile/library/common/internal_engine.cc
Comment (fredyw) [reply mobile/library/common/internal_engine.cc:288]: Right, but the callback name is `onConnectionTypChanged` and not `onConnectionChanged`: https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=974;drc=8e948282d37c0e119e3102236878d6f4d5052c16

Especially for now we are only interested in the changes in the network type and not on all changes in the network capabilities, hence the rename. But if you feel strongly about it, I can change it back. Let me know.

**Review by abeyad — DISMISSED** [review]
LGTM but please also wait for Dan's approval, thanks for fixing this!

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (danzh2010) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: If onCapabilitiesChanged() is called before onAvailable() in versions under O, the current logic needs to happen under onAvailable(), right?  Otherwise, the DNS refreshment may still fail if the network is not available yet.

SDK 21 is added in version L which was before version O. To avoid the complexity, I would suggest to just support network change after version O.

File: mobile/library/common/internal_engine.cc
Comment (danzh2010) [reply mobile/library/common/internal_engine.cc:288]: We don't need to exactly follow the naming of Chromium code, especially when Chromium naming is confusing. If you don't feel like to change it back, I can do it in a follow up PR.

File: mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java
Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: I updated the PR description.

Comment (fredyw) [reply mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java]: I put the old comment back.

File: mobile/library/common/internal_engine.cc
Comment (fredyw) [reply mobile/library/common/internal_engine.cc]: Done. I reverted the rename.

**Review by danzh2010 — APPROVED** [review]
LGTM

**Review by abeyad — APPROVED** [review]

**Pull request merged**
